### ISSUE TITLE
registry resource is updated when using dword 4294967295 (ffffffff)

### ISSUE THREAD

**Issue by lbn-aaio-team** [chef/chef#5104]
## Description

At each start of chef-client on windows, the registry key resource, when a dword max size (4294967295) value is set, is always updating the resource despite that the value is correctly set in the windows registry.

## Chef Version

chef-client x64 for windows v12.12.15 (tested with 12.11.18 too)

## Platform Version

Windows 2012 R2 x64 En-US (on vmware)

## Replication Case

registry_key "HKLM\\SYSTEM\\CurrentControlSet\\Control\\SecurityProviders\\SCHANNEL\\Ciphers\\AES 256/256" do
  values  [{:name => 'Enabled', :type => :dword, :data => '4294967295'}]
  action  :create
  recursive  true
end

Run first chef-client : the registry value is successfuly updated.
Run second chef-client command: the regitry value is updated again
The value "Enabled" is applied at each chef-client run.

## Client Output

Starting Chef Client, version 12.12.15
[2016-07-12T11:54:55+02:00] INFO: *** Chef 12.12.15 ***
[2016-07-12T11:54:55+02:00] INFO: Platform: x64-mingw32
[...]
Converging 1 resources
Recipe: win-security::security_providers
  * registry_key[HKLM\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Ciphers\AES 256/256] action create[201
6-07-12T11:55:05+02:00] INFO: Processing registry_key[HKLM\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Ciphers\AES 256/256] action create (win-security::security_providers line 1)
    - set value {:name=>"Enabled", :type=>:dword, :data=>"4294967295"}
[2016-07-12T11:55:06+02:00] INFO: Chef Run complete in 2.500202 seconds

Running handlers:
[2016-07-12T11:55:06+02:00] INFO: Running report handlers
[2016-07-12T11:55:07+02:00] INFO: Sending attributes to foreman
  - ChefHandlerForeman::ForemanFacts
Running handlers complete
[2016-07-12T11:55:07+02:00] INFO: Report handlers complete
[...]
Chef Client finished, 1/1 resources updated in 12 seconds

## Client full trace
[trace.txt](https://github.com/chef/chef/files/359025/trace.txt)

**Comment by lamont-granquist** [comment]
i feel like i've read a similar bug recently, might be a dup?

**Comment by lbn-aaio-team** [comment]
Hello,

I don't understand why this issue is low priority. Have you an idea of the chef-client release which fix this bug?
Actually, our reporting tools is still contaminated by all changes applied at each chef-client run on all Windows servers. 
If the problem is not fixed, we should probably set up a guard to make the change be skipped. But it is not the way we like to use chef ...

Regards,

**Comment by dheerajd-msys** [comment]
Hi @lbn-aaio-team ,

After doing some analysis, found that chef has Recipe DSL methods like `registry_key_exists?,registry_data_exists?`
Use `registry_data_exists?` method  in `not_if` block of the `registry_key` resource, so if the data is already set correctly, it won't update it.
More details on the usage of the DSL methods can be found [here](https://docs.chef.io/resource_registry_key.html#recipe-dsl-methods)

Let me know if you face any issues.

**Comment by dheerajd-msys** [comment]
Hi @lbn-aaio-team, 

My further analysis is as follows:-
In the chef code mentioned [here](https://github.com/chef/chef/blob/master/lib/chef/provider/registry_key.rb#L105), it checks if the value which needs to be set is same as that which is currently set. 
```
current_value = {:name => 'Enabled', :type => :dword, :data => 4294967295}   #current value set in registry key
value = {:name => 'Enabled', :type => :dword, :data => "4294967295"}         #value passed by user which needs to be set
```

The issue occurs for type `dword`, as when the data value passed as string in registry_key resource,   however while actually setting the data value in registry, it is set as Fixnum datatype(Integer).
During second chef client run, after verifying that registry key exist, during comparison of the set data and the new data value to be set, the values differ as the comparison happens for `integer(current_value[:data])` and `string(value[:data])` 

The workaround which works is in the data value of the registry_key resource, instead of assigning string value to data, assign it as integer value as follows.

```
registry_key "HKLM\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Ciphers\AES 256/256" do
values [{:name => 'Enabled', :type => :dword, :data => 4294967295}]
action :create
recursive true
end
```

**Comment by lbn-aaio-team** [comment]
Hi @dheerajd-msys,

Thank you very much for your help. 
I deploy my cookbook with your method and all is working fine now.
This is more perform than a powershell guard :)

**Comment by btm** [comment]
@dheerajd-msys when type is dword, then data should always be a 32-bit number, yes? In that case we should patch chef to coerce the data to an integer so users don't need to learn about this caveat.

**Comment by btm** [comment]
This is automatically fixed now by #5268, which will be in an upcoming release (probably october).
